junit-mock/src/lib.rs

```diff
@@ -0,0 +1,452 @@
+    #[arg(long, conflicts_with_all = ["test_case_names", "test_case_classnames"], default_value = "10")]
+    pub test_case_random_count: usize,
+
+    /// The chance of a system out message being added to the test case
+    #[arg(long, value_parser = clap::value_parser!(u8).range(0..=100), default_value = "50")]
+    pub test_case_sys_out_percentage: u8,
+
+    /// The chance of a system error message being added to the test case
+    #[arg(long, value_parser = clap::value_parser!(u8).range(0..=100), default_value = "50")]
+    pub test_case_sys_err_percentage: u8,
+
+    /// Inclusive range of time between test case timestamps
+    #[arg(long, num_args = 1..=2, value_names = ["DURATION_RANGE_START", "DURATION_RANGE_END"], default_values = ["30s", "1m"])]
+    pub test_case_duration_range: Vec<humantime::Duration>,
+
+    /// The chance of a test case succeeding, skipping, failing, and erroring (must add up to 100)
+    #[arg(long, value_parser = four_percentages_parser, default_value = "25,25,25,25")]
+    pub test_case_success_to_skip_to_fail_to_error_percentage: Vec<Vec<u8>>,
+}
+
+percentages_parser!(two_percentages_parser, 2);
+
+#[derive(Debug, Parser)]
+#[group()]
+pub struct TestRerunOptions {
+    /// Inclusive range of the number of reruns of a test that was not skipped
+    #[arg(long, num_args = 1..=2, value_names = ["COUNT_RANGE_START", "COUNT_RANGE_END"], default_values = ["0", "2"])]
+    pub test_rerun_count_range: Vec<usize>,
+
+    /// The chance of a test rerun failing and erroring (must add up to 100)
+    #[arg(long, value_parser = two_percentages_parser, default_value = "50,50")]
+    pub test_rerun_fail_to_error_percentage: Vec<Vec<u8>>,
+}
+
+pub struct JunitMock {
+    seed: u64,
+    rng: StdRng,
+    options: Options,
+
+    // state for generating reports
+    timestamp: chrono::DateTime<chrono::Utc>,
+    total_duration: Duration,
+}
+
+impl JunitMock {
+    pub fn new(options: Options) -> Self {
+        let (seed, rng) = JunitMock::rng_from_seed(&options);
+        Self {
+            seed,
+            rng,
+            options,
+            timestamp: chrono::Utc::now(),
+            total_duration: Duration::new(0, 0),
+        }
+    }
+
+    fn rng_from_seed(options: &Options) -> (u64, StdRng) {
+        let seed = options.global.seed.unwrap_or_else(rand::random);
+        (seed, StdRng::seed_from_u64(seed))
+    }
+
+    pub fn set_options(&mut self, options: Options) {
+        let (seed, rng) = JunitMock::rng_from_seed(&options);
+        self.seed = seed;
+        self.rng = rng;
+        self.options = options;
+    }
+
+    pub fn get_seed(&self) -> u64 {
+        self.seed
+    }
+
+    pub fn increment_duration(&mut self, duration: Duration) {
+        self.total_duration += duration;
+        self.timestamp += duration;
+    }
+
+    pub fn generate_reports(&mut self) -> Vec<Report> {
+        self.timestamp = chrono::Utc::now();
```

Comment:
is `timestamp` something you should also potentially take as an option? then you can have fully deterministic outputs if someone provides the same seed and `start timestamp` then I assume the output should be identical

Reply:
We can do that. I thought about it, but then I figured it'd be a weird / possibly breaking if our timestamps were different than the modified date, but now that I think about it I made the timestamps go into the future... So, I suppose we should accept an initial timestamp!